common/core/hwclayerbuffermanager.cpp

```diff
@@ -0,0 +1,88 @@
+/*
+// Copyright (c) 2016 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+*/
+
+#include "hwclayerbuffermanager.h"
+
+namespace hwcomposer {
+
+HwcLayerBufferManager::HwcLayerBufferManager() {
+  for(size_t i = 0; i < BUFFER_CACHE_LENGTH; i++)
+    cached_buffers_.emplace_back();
+}
+
+HwcLayerBufferManager::~HwcLayerBufferManager() {
+    PurgeBuffer();
+}
+
+void HwcLayerBufferManager::PurgeBuffer() {
+  for(auto &map : cached_buffers_) {
+  	map.clear();
+  }
+}
+
+void HwcLayerBufferManager::PurgeGraphicsResources() {
+  for (auto& map : cached_buffers_) {
+    for (auto& pair : map) {
+      pair.second->DeleteImage();
+      pair.second->DeleteTexture();
+    }
+  }
+}
+
+void HwcLayerBufferManager::Dump() {
+}
+
+std::shared_ptr<OverlayBuffer>& HwcLayerBufferManager::FindCachedBuffer(const HWCNativeBuffer& native_buffer) {
+
+  BUFFER_MAP&  first_map = cached_buffers_[0];
+  static std::shared_ptr<OverlayBuffer> pBufNull = nullptr;
+  for (auto& map : cached_buffers_) {
+    if (map.count(native_buffer)) {
+      std::shared_ptr<OverlayBuffer>& pBuf = map[native_buffer];
+      if (&map != &first_map) {
+        first_map.emplace(std::make_pair(native_buffer, pBuf));
+        map.erase(native_buffer);
+      }
+#ifdef CACHE_TRACING
+      hit_count_++;
+#endif
+      return pBuf;
+    }
+  }
+
+#ifdef CACHE_TRACING
+  miss_count_++;
+  if (miss_count_ % 100 == 0)
+    ICACHETRACE("cache miss count is %llu, while hit count is %llu",
+                miss_count_, hit_count_);
+#endif
+
+  return pBufNull;
+}
+
+void HwcLayerBufferManager::RegisterBuffer(const HWCNativeBuffer& native_buffer, std::shared_ptr<OverlayBuffer>& pBuffer) {
+
+  BUFFER_MAP&  first_map = cached_buffers_[0];
+  first_map.emplace(std::make_pair(native_buffer, pBuffer));
+}
+
+void HwcLayerBufferManager::RefreshBufferCache() {
+  auto begin = cached_buffers_.begin();
+  cached_buffers_.emplace(begin);
+  cached_buffers_.pop_back();
```

Comment:
This line:  cached_buffers_.pop_back();  would destroy buffers right ? In this case you want to move it to another list and delete these buffers from Compositor thread.

Reply:
Thinking more on this case, simple way would be store these buffers which go out of scope in a separate vector, in DisplayQueue after u do the current commit, query for these buffers from BufferManager. If vector is not empty, pass the ownership of these buffers to DisplayQueue. DisplayQueue would than post a task to CompositorThread. Compositor thread in its side would delete all the buffers which dont need gl context first (bu checking if buffer has valid gl resources). If it has buffers which have gl resources, than it will ensure context is initialized and delete these buffers.

Reply:
Actually I am bit confused what this function is also doing. Can you explain ?

Reply:
@js0701 On top of this work I refactored the internals so all resources are only released in compositorthread:
https://github.com/intel/IA-Hardware-Composer/commits/next

Can you test this on u side and see if surfacedamage still works like before ? If things look good I can send it as PR for review.

Reply:
Also, I see lot of these fb creation failures during test commit:
"drmModeAddFB2 error (1920x1280, AB24, handle 9 pitch 7680) (No such file or directory)"

I think we want to understand this before moving next to master branch. This happens with Android IA, when u use launcher to switch between all opened applications. @js0701  Do you have env setup or @harishkrupo Can you help ?

Reply:
really?  I will try in my env and investigate why. I have 2 days trainning and may have some delays